web: migrate part of repo container routes

valerybugakov opened the pull request:
## Context

RR6 migration for the `RepoContainer` and `RepoRevisionContainer` routes. Also, removed a lot of dead code.
Part of #33834 

## Test plan

1. CI main-dry-run [here](https://github.com/sourcegraph/sourcegraph/pull/47014/commits/b75d5b9f67f8d174c05c8b1cf28e98d0946f2c8e).
2. Manually visited all routes locally.

## App preview:

- [Web](/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| -0.06% (-1.60 kb) | -0.15% (-21.75 kb) 🔽 | -0.17% (-20.14 kb) 🔽 | -0.14% (-1) 🔽 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/vb/rr6-migration-2/compare-report.html;tab=live_object) for a full comparison between the commits 59eac83 and a4f1797 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

valerybugakov commented on client/web/src/repo/blob/Blob.tsx:
```diff
@@ -115,6 +114,19 @@ export interface BlobProps
+
+    navigate: NavigateFunction
+
+    /**
+     * TODO(valery): RR6
+     * @deprecated prefer using useNavigate()
+     */
+    history: H.History
+    /**
+     * TODO(valery): RR6
+     * @deprecated prefer using useLocation()
+     */
+    location: Location
```
Preserve these props for now because multiple CodeMirror modules should be updated after their deletion.

valerybugakov commented on client/web/src/repo/commits/GitCommitNode.tsx:
```diff
@@ -238,29 +234,6 @@ export const GitCommitNode: React.FunctionComponent<React.PropsWithChildren<GitC
         </Code>
     )
 
-    if (sidebar) {
```
Dead code

philipp-spiess commented on client/web/src/SourcegraphWebApp.tsx:
```diff
@@ -404,17 +360,47 @@ export class SourcegraphWebApp extends React.Component<
                 ]}
             >
-                <Router history={history} key={0}>
+                <Router history={globalHistory}>
```
😅

philipp-spiess commented on client/web/src/components/RedirectRoute.tsx:
```diff
@@ -0,0 +1,16 @@
+    const location = useLocation()
+    const params = useParams()
+
+    return <Navigate to={getRedirectURL({ location, params })} />
```
Redirects should replace, no? Otherwise the back button will just redirect again :D

```suggestion
    return <Navigate replace={true} to={getRedirectURL({ location, params })} />
```

philipp-spiess commented on client/web/src/components/diff/FileDiffNode.tsx:
```diff
@@ -33,6 +34,7 @@ export const FileDiffNode: React.FunctionComponent<React.PropsWithChildren<FileD
     persistLines,
     diffMode = 'unified',
 }) => {
+    const location = useLocation()
```
I noticed this in a lot of places that we use `window.location` over the RR location. 😢

philipp-spiess commented on client/web/src/enterprise/batches/repo/RepositoryBatchChangesArea.tsx:
```diff
@@ -1,50 +1,35 @@
-    globbing: boolean
 }
 
+const BREADCRUMB = { key: 'batch-changes', element: 'Batch Changes' }
```
noice

philipp-spiess commented on client/web/src/enterprise/batches/repo/RepositoryBatchChangesArea.tsx:
```diff
@@ -1,50 +1,35 @@
 
     return (
         <div className="repository-batch-changes-area container mt-3">
-            <Switch>
-                <Route
-                    path={`${match.url}`}
-                    key="hardcoded-key" // see https://github.com/ReactTraining/react-router/issues/4578#issuecomment-334489490
-                    exact={true}
-                    render={routeComponentProps => <BatchChangeRepoPage {...routeComponentProps} {...props} />}
-                />
-                <Route key="hardcoded-key" component={NotFoundPage} />
-            </Switch>
```
What the

philipp-spiess commented on client/web/src/enterprise/repo/enterpriseRepoContainerRoutes.tsx:
```diff
@@ -21,12 +19,14 @@ export const enterpriseRepoContainerRoutes: readonly RepoContainerRoute[] = [
     {
-        path: '/-/code-graph',
-        exact: false,
+        path: '/-/code-graph/*',
```
Do we also need to match `/-/codegraph` (without sub-routes?) If so you need to duplicate the rule into a second one that just uses `/-/code-graph` as the `path`.

valerybugakov replied:
Verified that it works as is and replaced `Redirect` with `Navigate` there

philipp-spiess commented on client/web/src/enterprise/repo/enterpriseRepoContainerRoutes.tsx:
```diff
@@ -21,12 +19,14 @@ export const enterpriseRepoContainerRoutes: readonly RepoContainerRoute[] = [
     {
         path: '/-/code-intelligence',
-        exact: false,
-        render: props => <Redirect to={props.location.pathname.replace('/code-intelligence', '/code-graph')} />,
+        render: () => (
+            <RedirectRoute
+                getRedirectURL={({ location }) => location.pathname.replace('/code-intelligence', '/code-graph')}
+            />
+        ),
```
This needs a second rule with the same render function and `/-/code-intelligence/*` as a route

philipp-spiess commented on client/web/src/enterprise/repo/settings/routes.tsx:
```diff
@@ -16,23 +15,16 @@ export const enterpriseRepoSettingsAreaRoutes: readonly RepoSettingsAreaRoute[]
     {
         path: '/code-intelligence',
-        exact: false,
-        render: props => <Redirect to={props.location.pathname.replace('/settings/', '/')} />,
+        render: () => <RedirectRoute getRedirectURL={({ location }) => location.pathname.replace('/settings/', '/')} />,
     },
```
This also needs a second route for `/code-intelligence/*`

philipp-spiess commented on client/web/src/repo/RepositoryFileTreePage.tsx:
```diff
@@ -20,35 +20,27 @@ import { TreePage } from './tree/TreePage'
+    const location = useLocation()
+    const { filePath = '' } = parseBrowserRepoURL(location.pathname)
 
-    // The decoding depends on the pinned `history` version.
-    // See https://github.com/sourcegraph/sourcegraph/issues/4408
-    // and https://github.com/ReactTraining/history/issues/505
-    const filePath = decodeURIComponent(match.params.filePath || '') // empty string is root
```
Do we no longer need to do the decoding here?

valerybugakov replied:
Good catch! It was not necessary with useParams because RR decodes them for us. It's needed now because we use `parseBrowserRepoURL`.

philipp-spiess commented on client/web/src/repo/RepositoryFileTreePage.tsx:
```diff
@@ -60,7 +52,7 @@ export const RepositoryFileTreePage: React.FunctionComponent<
             `L${startLineNumber}` + (endLineNumber ? `-${endLineNumber}` : '')
         )
-        return <Redirect to={url} />
+        return <Navigate to={url} />
```
replace?

philipp-spiess commented on client/web/src/repo/RepositoryFileTreePage.tsx:
```diff
@@ -20,35 +20,27 @@ import { TreePage } from './tree/TreePage'
-    if (match.params.objectType && filePath.replace(/\/+$/g, '') === '') {
-        return <Redirect to={toRepoURL({ repoName, revision: context.revision })} />
+    if (maybeObjectType && filePath.replace(/\/+$/g, '') === '') {
+        return <Navigate to={toRepoURL({ repoName, revision: context.revision })} />
```
Should we use replace here?

philipp-spiess commented on client/web/src/repo/RepositoryFileTreePage.tsx:
```diff
@@ -73,14 +65,13 @@ export const RepositoryFileTreePage: React.FunctionComponent<
         const range = formatLineOrPositionOrRange(parsedQuery)
         const url = appendLineRangeQueryParameter(location.pathname + location.search, range ? `L${range}` : undefined)
-        return <Redirect to={url + formatHash(hashParameters)} />
+        return <Navigate to={url + formatHash(hashParameters)} />
```
replace?

philipp-spiess commented on client/web/src/repo/actions/GoToPermalinkAction.tsx:
```diff
@@ -13,97 +13,72 @@ import { replaceRevisionInURL } from '../../util/url'
-         * The commit SHA for the revision in the current location (URL).
-         */
-        commitID: string
+export const GoToPermalinkAction: React.FunctionComponent<GoToPermalinkActionProps> = props => {
```
ty i was too lazy to migrate any of the class components yet haha

philipp-spiess commented on client/web/src/repo/blob/BlobPage.tsx:
```diff
@@ -446,7 +442,7 @@ export const BlobPage: React.FunctionComponent<React.PropsWithChildren<BlobPageP
         // We don't have error names on GraphQL errors.
         if (/not a blob/i.test(blobInfoOrError.message)) {
-            return <Redirect to={toTreeURL(props)} />
+            return <Navigate to={toTreeURL(props)} />
```
replace?

valerybugakov commented on client/web/src/extensions/components/ActionItemsBar.story.tsx:
```diff
@@ -102,7 +93,6 @@ export const Default: React.FunctionComponent<React.PropsWithChildren<unknown>>
     return (
         <ActionItemsBar
             repo={undefined}
-            location={LOCATION}
```
Need to update the story to provide location via react-router context.

philipp-spiess commented on client/web/src/repo/commit/RepositoryCommitPage.tsx:
```diff
@@ -58,10 +53,16 @@ export type { DiffMode } from '@sourcegraph/shared/src/settings/temporary/diffMo
 export const RepositoryCommitPage: React.FunctionComponent<RepositoryCommitPageProps> = props => {
+    const params = useParams<{ revspec: string }>()
+
+    if (!params.revspec) {
+        throw new Error('Missing `revspec` param!')
+    }
```
Won't this complain about RuLeS oF HoOks?

valerybugakov replied:
The ESLint is happy about it 🤷‍♂️

philipp-spiess commented on client/web/src/repo/compare/RepositoryCompareArea.tsx:
```diff
@@ -76,30 +63,21 @@ export const RepositoryCompareArea: React.FunctionComponent<RepositoryCompareAre
             {spec === null ? (
                 <Alert variant="danger">Invalid comparison specifier</Alert>
             ) : (
-                <Switch>
-                    <Route
-                        path={`${match.url}`}
-                        key="hardcoded-key" // see https://github.com/ReactTraining/react-router/issues/4578#issuecomment-334489490
-                        exact={true}
-                        render={routeComponentProps => (
-                            <RepositoryCompareOverviewPage
-                                {...routeComponentProps}
-                                {...commonProps}
-                                path={path}
-                                isLightTheme={isLightTheme}
-                            />
-                        )}
-                    />
-                    <Route key="hardcoded-key" component={NotFoundPage} />
-                </Switch>
```
What's up with this pattern why is this happening the second time here? 😕

valerybugakov replied:
I think it's because of the "convention". Area components all have routes inside of them. But I think it makes sense to remove redundant switches for out migration. We will probably want to lift these routes higher in the tree to simplify the structure.

philipp-spiess commented on client/web/src/repo/tree/TreePage.tsx:
```diff
@@ -305,7 +300,7 @@ export const TreePage: React.FunctionComponent<React.PropsWithChildren<Props>> =
                     // We don't have error names on GraphQL errors.
                     /not a directory/i.test(treeOrError.message) ? (
-                        <Redirect to={toPrettyBlobURL({ repoName, revision, commitID, filePath })} />
+                        <Navigate to={toPrettyBlobURL({ repoName, revision, commitID, filePath })} />
```
replace?

philipp-spiess commented on client/web/src/util/contributions.ts:
```diff
@@ -32,6 +32,17 @@ export interface RouteDescriptor<C extends object = {}, P extends object = any>
+ *
+ * @template C Context information that is passed to `render` and `condition`
+ */
+export interface RouteV6Descriptor<C extends object = {}> extends Conditional<C> {
```
Ahaha I called it RouteDescriptorV6 in https://github.com/sourcegraph/sourcegraph/pull/47276

valerybugakov replied:
Great minds...

philipp-spiess reviewed (approved):
Amazing work! 

![Omg Clap](https://media.giphy.com/media/xT77XWum9yH7zNkFW0/giphy.gif)

umpox commented on client/web/webpack.config.js:
```diff
@@ -166,17 +166,16 @@ const config = {
-        filter: ({ isInitial, name }) =>
-          isInitial || Object.values(initialChunkNames).some(initialChunkName => name?.includes(initialChunkName)),
-      }),
+    new WebpackManifestPlugin({
```
Is this diff meant to be in this PR? Feels detached from everything else (fine if valid reason just checking)

valerybugakov replied:
Thanks for checking! We rely on Webpack manifest for our integration tests, and I noticed that if I want to use the standalone server for test debugging, we don't flush the manifest to the disk. This change fixes that.

umpox reviewed (commented):
Nice! Left some comments below

umpox commented:
Noticed a couple of weird things when testing around, not sure if entirely related to this PR or not, end of my day now but can follow up tomorrow:

1. URL hash behavior seems off. Select the same commit as the one `main` is on right now, it'll remove that commit from the URL when any file is clicked. Feels unexpected if I want to keep that hash intact. Also when clicking through the sidebar it seems the URL changes often between the branch name and the current commit. 

2. May be unrelated but this URL: https://sourcegraph.test:3443/github.com/sourcegraph/sourcegraph@558867528e2c07a87d22e4434df141653f98e55d/-/blob/client/codeintellify/src/hoverifier.test.ts completely breaks for me with 
![image](https://user-images.githubusercontent.com/9516420/216387871-01115fd6-997c-46f6-ad2b-2c3a1759f538.png)

vovakulikov commented on client/web/src/enterprise/batches/repo/RepositoryBatchChangesArea.tsx:
```diff
@@ -1,50 +1,35 @@
+export const RepositoryBatchChangesArea: FC<RepositoryBatchChangesAreaPageProps> = props => {
+    const { useBreadcrumb, repo, isLightTheme } = props
+
+    useBreadcrumb(BREADCRUMB)
```
Passing hooks through props, it scares me every time I see this

valerybugakov replied:
Agree, we probably could extract the callback from the context instead of passing it down like this.

vovakulikov commented on client/web/src/enterprise/codeintel/configuration/pages/CodeIntelConfigurationPage.tsx:
```diff
@@ -107,13 +105,17 @@ export const CodeIntelConfigurationPage: FunctionComponent<CodeIntelConfiguratio
+                        pathname: './configuration',
+                    },
+                    {
+                        state: { modal: 'SUCCESS', message: `Configuration policy ${name} has been deleted.` },
```
Haven't seen that before that we use location state as a common state store.

vovakulikov reviewed (approved):
LGTM, Tested this manually. One thing that I noticed but couldn't find this in this PR diff (so probably this is something else) is in the blob view on k8s, we render links for the go to definition and find references button, but in this PR, we render buttons (and if I click the button with CMD click this doesn't open link in the new tab but in the current tab)

I think I've seen something similar before from unknwon in slack

vovakulikov commented:
Ah, I actually missed the git blame panel. If you go to this [file](https://sourcegraph.test:3443/github.com/sourcegraph/sourcegraph@main/-/blob/client/codeintellify/src/hoverifier.test.ts) and open the git-blame panel then 

<img width="1096" alt="Screenshot 2023-02-02 at 23 21 36" src="https://user-images.githubusercontent.com/18492575/216497154-538f6145-517d-46bc-ba14-33098d9d0f26.png">

valerybugakov commented on client/web/src/components/HeroPage.tsx:
```diff
@@ -51,3 +52,11 @@ export const HeroPage: React.FunctionComponent<React.PropsWithChildren<HeroPageP
+    <HeroPage
+        icon={MapSearchIcon}
+        title="404: Not Found"
+        subtitle="Sorry, the requested repository tags page was not found."
```
Fixed it to be specific for all pages via prop.

valerybugakov commented on client/web/src/enterprise/codeintel/repo/RepositoryCodeIntelArea.tsx:
```diff
@@ -60,69 +60,51 @@ const CodeIntelConfigurationPolicyPage = lazyComponent<
     },
 ]
 
-const NotFoundPage: React.FunctionComponent<React.PropsWithChildren<unknown>> = () => (
-    <HeroPage
-        icon={MapSearchIcon}
-        title="404: Not Found"
-        subtitle="Sorry, the requested repository page was not found."
-    />
-)
```
I preserved the unique page types per message. We can still use `HeroPage` directly if we want to provide more info; I just noticed that `NotFoundPage` is copy-pasted everywhere with almost no changes and sometimes without a subtitle.

valerybugakov commented on client/web/src/enterprise/site-admin/routes.tsx:
```diff
@@ -102,7 +102,7 @@ export const enterpriseSiteAdminAreaRoutes: readonly SiteAdminAreaRoute[] = (
         {
-            path: '/code-intelligence',
+            path: '/code-intelligence/*',
             exact: false,
```
This path should not be changed

valerybugakov commented:
Thanks, @vovakulikov! I fixed the issue with git blame. I wasn't able to repro the problem related to the references button. 
Thanks, @umpox! I addressed your comments but cannot repro the issues you mentioned. I suspect they were fixed by other changes I made before your review.

Numbers88s commented on client/web/src/enterprise/repo/routes.tsx:
```diff
@@ -1,42 +0,0 @@
-    {
-        path: '/-/code-intelligence',
-        exact: false,
-        render: props => <Redirect to={props.location.pathname.replace('/code-intelligence', '/code-graph')} />,
```
Found it.
